In [25]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import plotly.express as px

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.datasets import load_iris

import tensorflow as tf
from tensorflow import keras

import warnings
warnings.filterwarnings("ignore")

In [26]:
# load data

df_iris_sklearn = load_iris()

X = pd.DataFrame(data = df_iris_sklearn.data,
                 columns = df_iris_sklearn.feature_names)

y = df_iris_sklearn.target

display(X.head(), y[:5])

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

array([0, 0, 0, 0, 0])

In [27]:
# split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, 
                                                    random_state = 42, 
                                                    stratify = y, 
                                                    shuffle = True)

# standardize the dataset
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

# build model
model = keras.Sequential([
    keras.layers.Dense(10, input_dim = 4, activation = 'relu', kernel_initializer = 'he_normal', 
                       kernel_regularizer = keras.regularizers.l2(0.01)),
    keras.layers.BatchNormalization(),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(7, activation = 'relu', kernel_initializer = 'he_normal', 
                       kernel_regularizer = keras.regularizers.l1_l2(l1 = 0.001, l2 = 0.001)),
    keras.layers.BatchNormalization(),
    keras.layers.Dropout(0.3),
    keras.layers.Dense(5, activation = 'relu', kernel_initializer = 'he_normal', 
                       kernel_regularizer = keras.regularizers.l1_l2(l1 = 0.001, l2 = 0.001)),
    keras.layers.Dense(3, activation = 'softmax')
])

# compile model
model.compile(optimizer = tf.optimizers.Adam(),
              loss = 'sparse_categorical_crossentropy',
              metrics = ['accuracy'])

In [28]:
model.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_16 (Dense)                │ (None, 10)             │            50 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_8           │ (None, 10)             │            40 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_8 (Dropout)             │ (None, 10)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_17 (Dense)                │ (None, 7)              │            77 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_9           │ (None, 7)              │            28 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_9 (Dropout)             │ (None, 7)              │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_18 (Dense)                │ (None, 5)              │            40 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_19 (Dense)                │ (None, 3)              │            18 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 253 (1012.00 B)

 Trainable params: 219 (876.00 B)

 Non-trainable params: 34 (136.00 B)

In [20]:
# fit model
history = model.fit(X_train_scaled, y_train, batch_size = 16, epochs = 500)
history

Epoch 1/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.4832 - loss: 1.3849
Epoch 2/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.3803 - loss: 1.4264
Epoch 3/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.4456 - loss: 1.3454
Epoch 4/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.5430 - loss: 1.3679
Epoch 5/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.4669 - loss: 1.3191
Epoch 6/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4651 - loss: 1.3552
Epoch 7/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5734 - loss: 1.2487
Epoch 8/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5193 - loss: 1.1982
Epoch 9/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4805 - loss: 1.2667
Epoch 10/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 976us/step - accuracy: 0.6253 - loss: 1.1811
Epoch 11/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.4767 - loss: 1.1368
Epoch 12/500
8/8 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.5396 - loss: 1.1336

In [21]:
# evaluate model
model.evaluate(X_test_scaled, y_test)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 137ms/step - accuracy: 0.9667 - loss: 0.1202


[0.12018550932407379, 0.9666666388511658]

In [22]:
def show_train_summary(model_name, history, epochs):
    """
    Visualizes the training summary for a model's performance over epochs using an interactive scatter plot.

    Args:
        model_name (str): The name of the model, used in the plot title.
        history (keras.callbacks.History): The history object returned by model.fit(), containing training metrics.
        epochs (list): A list of epoch numbers, typically ranging from 1 to the total number of epochs.

    Returns:
        None: Displays the plot in an interactive format.
    """
    # Convert training history to a DataFrame for easier manipulation
    history_dict = history.history
    data = {
        'Epoch': epochs,
        'Train Accuracy': history_dict["accuracy"],
        'Train Loss': history_dict["loss"]
    }

    df = pd.DataFrame(data)

    # Melt the DataFrame to make it suitable for Plotly Express
    df_melted = df.melt(id_vars = 'Epoch', value_vars = ['Train Accuracy', 'Train Loss'], 
                        var_name = 'Metric', value_name = 'Value')

    # Create the scatter plot using px.scatter
    fig = px.scatter(df_melted, x = 'Epoch', y = 'Value', color = 'Metric',
                     title = f"Model {model_name} Training Summary",
                     labels = {'Value': 'Value', 'Metric': 'Metric'})

    # Customize the layout
    fig.update_layout(
        height = 600,
        width = 1100,
        title_x = 0.5,
        title_xanchor = 'center',
        legend = dict(
            orientation = 'h',
            yanchor = 'bottom',
            y = -0.2,
            xanchor = 'center',
            x = 0.5
        )
    )

    # Show the plot
    fig.show()

show_train_summary('Dense', history, list(range(1, 500 + 1)))

Training Accuracy: The accuracy starts at a lower value but quickly rises within the first 100 epochs, showing a rapid learning phase. After reaching approximately 0.9, the accuracy stabilizes and continues to oscillate slightly around this value. This suggests that the model is approaching its optimal accuracy level and is well-tuned for the dataset.

Training Loss: The loss begins at a high value and decreases quickly during the initial epochs, indicating effective learning. It continues to decline gradually, with minimal fluctuation around a low value as training progresses, suggesting good model convergence.

Overall, the model shows a good fit.

In [23]:
# predict test data
y_pred     = model.predict(X_test_scaled)
y_pred_max = np.argmax(y_pred, axis = 1)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 93ms/step


In [24]:
# create classification report
report = classification_report(y_test, y_pred_max)

print(report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       0.91      1.00      0.95        10
           2       1.00      0.90      0.95        10

    accuracy                           0.97        30
   macro avg       0.97      0.97      0.97        30
weighted avg       0.97      0.97      0.97        30



The classification report shows that the model performs well, with an overall accuracy of 97% on the test set.

- Class 0: Achieves a perfect precision, recall, and F1-score of 1.00, indicating no misclassifications for this class.
- Class 1: Has a precision of 0.91 and recall of 1.00, leading to an F1-score of 0.95. This suggests that all instances of Class 1 were correctly identified, but there were some false positives.
- Class 2: Shows a precision of 1.00 and recall of 0.90, also resulting in an F1-score of 0.95. This indicates that while there were no false positives, a few instances of Class 2 were missed.
The macro average and weighted average for precision, recall, and F1-score are both 0.97, reflecting consistent and balanced performance across all classes.